cmd/mt-kafka-mdm-report-out-of-order/inputooofinder.go

```diff
@@ -0,0 +1,193 @@
+	prefix        string
+	substr        string
+	doUnknownMP   bool
+
+	data        map[schema.MKey]Tracker
+	definitions map[schema.MKey]schema.MetricDefinition
+
+	groupByTag    string
+	groupedByName *map[string]int
+	groupedByTag  *map[string]int
+
+	lock sync.Mutex
+}
+
+func newInputOOOFinder(graceDuration time.Duration, prefix string, substr string, doUnknownMP bool, definitions map[schema.MKey]schema.MetricDefinition, groupByTag string, groupedByName *map[string]int, groupedByTag *map[string]int) *inputOOOFinder {
+	return &inputOOOFinder{
+		graceDuration,
+		prefix,
+		substr,
+		doUnknownMP,
+
+		make(map[schema.MKey]Tracker),
+		definitions,
+
+		groupByTag,
+		groupedByName,
+		groupedByTag,
+
+		sync.Mutex{},
+	}
+}
+
+func (ip *inputOOOFinder) ProcessMetricData(metric *schema.MetricData, partition int32) {
+	if ip.prefix != "" && !strings.HasPrefix(metric.Name, ip.prefix) {
+		return
+	}
+	if ip.substr != "" && !strings.Contains(metric.Name, ip.substr) {
+		return
+	}
+
+	mkey, err := schema.MKeyFromString(metric.Id)
+	if err != nil {
+		log.Errorf("could not parse id %q: %s", metric.Id, err.Error())
+		return
+	}
+
+	ip.lock.Lock()
+	defer ip.lock.Unlock()
+
+	// update index
+	_, ok := ip.definitions[mkey]
+	if !ok {
+		ip.definitions[mkey] = schema.MetricDefinition{
+			Name: metric.Name,
+			Tags: metric.Tags,
+		}
+	}
+
+	now := Msg{
+		Part: partition,
+		Seen: time.Now(),
+		Md:   *metric,
+	}
+	tracker, ok := ip.data[mkey]
+	if !ok {
+		ip.data[mkey] = Tracker{
+			Head: now,
+		}
+	} else {
+		if uint32(metric.Time)+uint32(ip.graceDuration.Seconds()) > tracker.Head.Time() {
+			tracker.Head = now
+			tracker.NumBad = 0
+			ip.data[mkey] = tracker
+		} else {
+			// if metric time + grace period <= head point time, update "bad", generate event and print
+			tracker.Bad = now
+			tracker.NumBad += 1
+			tracker.DeltaTime = tracker.Head.Time() - uint32(metric.Time)
+			tracker.DeltaSeen = uint32(now.Seen.Unix()) - uint32(tracker.Head.Seen.Unix())
+			ip.data[mkey] = tracker
+
+			// increment grouping counts
+			(*ip.groupedByName)[metric.Name]++
+			for _, tag := range metric.Tags {
+				kv := strings.Split(tag, "=")
+				if len(kv) != 2 {
+					log.Errorf("unexpected tag encoding %s", tag)
+					continue
+				}
+				if kv[0] == ip.groupByTag {
+					(*ip.groupedByTag)[kv[1]]++
+				}
+			}
+		}
+	}
+}
+
+func (ip *inputOOOFinder) ProcessMetricPoint(mp schema.MetricPoint, format msg.Format, partition int32) {
+	ip.lock.Lock()
+	defer ip.lock.Unlock()
+
+	metricDefinition, exists := ip.definitions[mp.MKey]
+	if !exists {
+		log.Errorf("metric definition not found")
+		return
+	}
+
+	if ip.prefix != "" && !strings.HasPrefix(metricDefinition.Name, ip.prefix) {
+		return
+	}
+	if ip.substr != "" && !strings.Contains(metricDefinition.Name, ip.substr) {
+		return
+	}
+
+	now := Msg{
+		Part: partition,
+		Seen: time.Now(),
+		Mp:   mp,
+	}
+	tracker, ok := ip.data[mp.MKey]
+	if !ok {
+		if !ip.doUnknownMP {
```

Comment:
Not sure this is the right place for this. Would this be if we failed to get the MetricDefinition?